src/algebraic_geometry/structure_sheaf.lean

```diff
@@ -0,0 +1,233 @@
+so we replace his circumlocution about functions into a disjoint union with
+`Π x : U, localizations x`.
+-/
+def is_locally_fraction
+  {U : opens (Top.of (prime_spectrum R))} (f : Π x : U, localizations R x) : Prop :=
+∀ x : U, ∃ (V) (m : x.1 ∈ V) (i : V ⟶ U),
+  ∃ (r s : R), ∀ y : V,
+  ¬ (s ∈ y.1.as_ideal) ∧
+    f (i y : U) * (localization.of _).to_map s = (localization.of _).to_map r
+
+variables (R)
+
+/--
+Show that `is_locally_fraction_pred` is a "local predicate":
+in fact it is just the sheafification of the "prelocal predicate" `is_fraction`.
+-/
+def is_locally_fraction_local : local_predicate (localizations R) :=
+(is_fraction_prelocal R).sheafify
+
+@[simp]
+lemma is_locally_fraction_local_pred
+  {U : opens (Top.of (prime_spectrum R))} (f : Π x : U, localizations R x) :
+  (is_locally_fraction_local R).pred f = is_locally_fraction f :=
+rfl
+
+/--
+The functions satisfying `is_locally_fraction` form a subring.
+-/
+def sections_subring (U : (opens (Top.of (prime_spectrum R)))ᵒᵖ) :
+  subring (Π x : unop U, localizations R x) :=
+{ carrier := { f | is_locally_fraction f },
+  zero_mem' :=
+  begin
+    refine λ x, ⟨unop U, x.2, 𝟙 _, 0, 1, λ y, ⟨_, _⟩⟩,
+    { rw ←ideal.ne_top_iff_one, exact y.1.is_prime.1, },
+    { simp, },
+  end,
+  one_mem' :=
+  begin
+    refine λ x, ⟨unop U, x.2, 𝟙 _, 1, 1, λ y, ⟨_, _⟩⟩,
+    { rw ←ideal.ne_top_iff_one, exact y.1.is_prime.1, },
+    { simp, },
+  end,
+  add_mem' :=
+  begin
+    intros a b ha hb x,
+    rcases ha x with ⟨Va, ma, ia, ra, sa, wa⟩,
+    rcases hb x with ⟨Vb, mb, ib, rb, sb, wb⟩,
+    refine ⟨Va ⊓ Vb, ⟨ma, mb⟩, opens.inf_le_left _ _ ≫ ia, ra * sb + rb * sa, sa * sb, _⟩,
+    intro y,
+    rcases wa (opens.inf_le_left _ _ y) with ⟨nma, wa⟩,
+    rcases wb (opens.inf_le_right _ _ y) with ⟨nmb, wb⟩,
+    fsplit,
+    { intro H, cases y.1.is_prime.mem_or_mem H; contradiction, },
+    { simp only [add_mul, ring_hom.map_add, pi.add_apply, ring_hom.map_mul],
+      erw [←wa, ←wb],
+      simp only [mul_assoc],
+      congr' 2,
+      rw [mul_comm], refl, }
+  end,
+  neg_mem' :=
+  begin
+    intros a ha x,
+    rcases ha x with ⟨V, m, i, r, s, w⟩,
+    refine ⟨V, m, i, -r, s, _⟩,
+    intro y,
+    rcases w y with ⟨nm, w⟩,
+    fsplit,
+    { exact nm, },
+    { simp only [ring_hom.map_neg, pi.neg_apply],
+      erw [←w],
+      simp only [neg_mul_eq_neg_mul_symm], }
+  end,
+  mul_mem' :=
+  begin
+    intros a b ha hb x,
+    rcases ha x with ⟨Va, ma, ia, ra, sa, wa⟩,
+    rcases hb x with ⟨Vb, mb, ib, rb, sb, wb⟩,
+    refine ⟨Va ⊓ Vb, ⟨ma, mb⟩, opens.inf_le_left _ _ ≫ ia, ra * rb, sa * sb, _⟩,
+    intro y,
+    rcases wa (opens.inf_le_left _ _ y) with ⟨nma, wa⟩,
+    rcases wb (opens.inf_le_right _ _ y) with ⟨nmb, wb⟩,
+    fsplit,
+    { intro H, cases y.1.is_prime.mem_or_mem H; contradiction, },
+    { simp only [pi.mul_apply, ring_hom.map_mul],
+      erw [←wa, ←wb],
+      simp only [mul_left_comm, mul_assoc, mul_comm],
+      refl, }
```

Comment:
This is essentially the same proof as add_mem' in some sense. But I don't really know if there's anything you can do about it.

Reply:
Yeah, I thought about proving it was a `submonoid` first, then hoping that `to_additive` might do `add_submonoid`, but the arguments are of course slightly different...